halo2_gadgets/src/ecc/chip/mul/incomplete.rs

```diff
@@ -12,22 +12,59 @@ use halo2_proofs::{
 
 use pasta_curves::{arithmetic::FieldExt, pallas};
 
+/// A helper struct for implementing single-row double-and-add using incomplete addition.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub(crate) struct DoubleAndAdd {
+    // x-coordinate of the accumulator in each double-and-add iteration.
+    pub(crate) x_a: Column<Advice>,
+    // x-coordinate of the point being added in each double-and-add iteration.
+    pub(crate) x_p: Column<Advice>,
+    // lambda1 in each double-and-add iteration.
+    pub(crate) lambda_1: Column<Advice>,
+    // lambda2 in each double-and-add iteration.
+    pub(crate) lambda_2: Column<Advice>,
+}
+
+impl DoubleAndAdd {
+    /// Derives the expression `x_r = lambda_1^2 - x_a - x_p`.
+    pub(crate) fn x_r(
+        &self,
+        meta: &mut VirtualCells<pallas::Base>,
+        rotation: Rotation,
+    ) -> Expression<pallas::Base> {
+        let x_a = meta.query_advice(self.x_a, rotation);
+        let x_p = meta.query_advice(self.x_p, rotation);
+        let lambda_1 = meta.query_advice(self.lambda_1, rotation);
+        lambda_1.square() - x_a - x_p
+    }
+
+    /// Derives the expression `Y_A = (lambda_1 + lambda_2) * (x_a - x_r)`.
```

Comment:
Without the factor of 2, this could consistently just be `y_a`. It's not critical though.